qiskit/transpiler/preset_passmanagers/level3.py

```diff
@@ -197,53 +198,57 @@ def _opt_control(property_set):
+            basis_gates,
+            approximation_degree,
+            unitary_synthesis_method,
+            unitary_synthesis_plugin_config,
+        )
     init.append(RemoveResetInZeroState())
     init.append(OptimizeSwapBeforeMeasure())
     init.append(RemoveDiagonalGatesBeforeMeasure())
     if coupling_map or initial_layout:
-        layout = PassManager()
-        layout.append(_given_layout)
-        layout.append(_choose_layout_0, condition=_choose_layout_condition)
-        layout.append(_choose_layout_1, condition=_vf2_match_not_found)
-        layout += common.generate_embed_passmanager(coupling_map)
-        vf2_call_limit = None
-        if pass_manager_config.layout_method is None and pass_manager_config.initial_layout is None:
-            vf2_call_limit = int(3e7)  # Set call limit to ~60 sec with retworkx 0.10.2
-        routing = common.generate_routing_passmanager(
-            routing_pass,
-            target,
-            coupling_map=coupling_map,
-            vf2_call_limit=vf2_call_limit,
-            backend_properties=backend_properties,
-            seed_transpiler=seed_transpiler,
-            use_barrier_before_measurement=not toqm_pass,
-        )
+        if layout_method not in {"trivial", "dense", "noise_adaptive", "sabre"}:
+            layout = plugin_manager.get_passmanager_stage(
+                "layout", layout_method, pass_manager_config
+            )
+        else:
+            layout = PassManager()
+            layout.append(_given_layout)
+            layout.append(_choose_layout_0, condition=_choose_layout_condition)
+            layout.append(_choose_layout_1, condition=_vf2_match_not_found)
+            layout += common.generate_embed_passmanager(coupling_map)
+        routing = routing_pm
     else:
         layout = None
         routing = None
-    translation = common.generate_translation_passmanager(
-        target,
-        basis_gates,
-        translation_method,
-        approximation_degree,
-        coupling_map,
-        backend_properties,
-        unitary_synthesis_method,
-        unitary_synthesis_plugin_config,
-    )
+    if translation_method not in {"translator", "synthesis", "unroller"}:
+        translation = plugin_manager.get_passmanager_stage(
+            "translation", translation_method, pass_manager_config
+        )
+    else:
+        translation = common.generate_translation_passmanager(
+            target,
+            basis_gates,
+            translation_method,
+            approximation_degree,
+            coupling_map,
+            backend_properties,
+            unitary_synthesis_method,
+            unitary_synthesis_plugin_config,
+        )
     pre_routing = None
     if toqm_pass:
         pre_routing = translation
-    optimization = PassManager()
-    unroll = [pass_ for x in translation.passes() for pass_ in x["passes"]]
-    optimization.append(_depth_check + _size_check)
+    optimization = None
+    if optimization_method is None:
```

Comment:
I dont fully get the difference between `optimization` and `optimization_method`. The usual pattern is
```
if X is None:
    <build X>
<use X>
```

Reply:
The level3 passmanager is a bit weird in that it builds the optimization stage with the pre_optimization stage based on a bunch of nested logic:

https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/transpiler/preset_passmanagers/level3.py#L244-L270

(also there might be a bug there it looks like it runs the optimization loop twice in that code)

I guess I got a bit overly eager with trying to deduplicate logic here. I'll expand this optimization section out in level 3 to make it easier to follow

Reply:
Fixed in https://github.com/Qiskit/qiskit-terra/pull/8305/commits/4190774c3b50a1a9b7832437dd4492f8ddea4bfc